File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/G._Сумма_четвёрок/solution.py
import sys
from collections import defaultdict


def solution(nums: list[int], target: int) -> None:
    """
    Time Complexity O(N**3)
    Memory Complexity O(N**4)
    """
    nums.sort()
    res: list[list[int]] = []
    n: int = len(nums)

    pair_sum: dict[int, list[tuple[int, int]]] = defaultdict(list)

    for i in range(n):
        for j in range(i + 1, n):
            pair_sum[nums[i] + nums[j]].append((i, j))

    for k in range(n):
        for l in range(k + 1, n):
            complement: int = target - (nums[k] + nums[l])
            if complement in pair_sum:
                for i, j in pair_sum[complement]:
                    if i < k and j < k:
                        res.append([nums[i], nums[j], nums[k], nums[l]])

    res: list[list[int]] = sorted(map(list, set(map(tuple, res))))

    sys.stdout.write(f'{len(res)}\n')
    for quadruplet in res:
        print(' '.join(str(num) for num in quadruplet))


if __name__ == '__main__':
    n: int = int(sys.stdin.readline().rstrip())
    target: int = int(sys.stdin.readline().rstrip())
    nums: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    solution(nums, target)

File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/G._Сумма_четвёрок/solution.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>

void solution(std::vector<long long>& nums, long long target) {
    /*
    Time Complexity O(N**3)
    Memory Complexity O(N**4)
    */
    std::sort(nums.begin(), nums.end());
    std::set<std::vector<long long>> result;
    int n = nums.size();

    for (int i = 0; i < n - 3; ++i) {
        if (i > 0 && nums[i] == nums[i - 1]) continue;
        for (int j = i + 1; j < n - 2; ++j) {
            if (j > i + 1 && nums[j] == nums[j - 1]) continue;
            int left = j + 1, right = n - 1;
            while (left < right) {
                long long current_sum = nums[i] + nums[j] + nums[left] + nums[right];
                if (current_sum < target) {
                    left++;
                } else if (current_sum > target) {
                    right--;
                } else {
                    result.insert({nums[i], nums[j], nums[left], nums[right]});
                    left++;
                    right--;
                    while (left < right && nums[left] == nums[left - 1]) left++;
                    while (left < right && nums[right] == nums[right + 1]) right--;
                }
            }
        }
    }

    std::cout << result.size() << "\n";
    for (const auto& quadruplet : result) {
        for (long long num : quadruplet) {
            std::cout << num << " ";
        }
        std::cout << "\n";
    }
}

int main() {
    int n;
    long long target;
    std::cin >> n >> target;
    std::vector<long long> nums(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> nums[i];
    }
    solution(nums, target);
    return 0;
}
